File: 338. Counting Bits.py
class Solution:
    def countBits(self, n: int) -> list[int]:
        if n == 0:
            return [0]
        if n == 1:
            return [0, 1]

        pivot = 1
        ans = [0] * (n + 1)
        ans[1] = 1
        i = 2
        while i < n + 1:
            if i / 2 == pivot:
                pivot = i
            ans[i] = 1 + ans[i - pivot]
            i += 1

        return ans


sol = Solution()
print(sol.countBits(5))

File: 338. Counting Bits.cpp
// 338. Counting Bits


class Solution {
public:
    vector<int> countBits(int n) {
        vector<int> res (n+1);

        if (n == 0) {
            return res;
        }
        res[0] = 0;
        res[1] = 1;
        if (n == 0) {
            return res;
        }

        int pivot = 1;
        for (int i = 2; i < n+1; i++) {
            if (i / 2 == pivot) {
                pivot = i; 
            }
            res[i] = 1 + res[i % pivot];

        }
        return res;
    }
};